### PR TITLE
add a close function and a visititems-function to request all datasets in a file

### PR DESCRIPTION
**Pull request by paulhfu**
Hi @constantinpape ,
Please let me know if you want anything included to group.close(). If it stays like that, it is sufficient for my purposes.

### CODE REVIEW

**Comment by constantinpape** [comment]
Thanks for the PR, looks good in general.
I will try to fix the travis builds later, so we have test coverage.

**Comment by constantinpape** [comment]
@paulhfu 
Actually, I have some comments on `close`.
If we want to keep it in line with the `h5py` API it should be only available for `File`, not for `Group`.
Also, in `h5py` `close` does not have a return value, so I would rather `pass` instead of `return True`.

Please make these changes unless you had a good reason for diverging from `h5py` here.

File: src/python/module/z5py/group.py [hunk 1]
```diff
@@ -228,3 +228,43 @@ def require_dataset(self, name, shape,
         return Dataset._require_dataset(path, shape, dtype, chunks,
                                         n_threads, self.is_zarr, self._internal_mode,
                                         **kwargs)
+
+    def visititems(self, func, path):
```
[the remark this answers is not in the document]
Comment (clbarnes) [reply src/python/module/z5py/group.py]: @constantinpape this raises another thought I've had which is to have some way of finding an object's internal path - i.e. the relative path to it from the nearest containing `File`. Or is this antithetical to N5's concept that it's all just `Group`s with an optional json file all the way up?

[the remark this answers is not in the document]
Comment (constantinpape) [reply src/python/module/z5py/group.py]: @clbarnes 
I think the internal path would be a nice feature to have.
As long as this is just part of the `z5py` API this doesn't go against the N5 concept.
This implies passing the internal path to the corresponding objects during construction though. 
Figuring out the internal path on the fly would imply making (potentially invalid) assumptions about the N5 file structure.

In this specific case, I think it should be fine to just leave out `path` as argument (which is def. necessary to make this function conform with `h5py`) and use `self.path`.

**Comment by constantinpape** [comment]
@clbarnes 
Thanks for having a closer look at this.
I agree with all your points and left some comment about the internal path.

@paulhfu 
Please also merge the current master to fix the travis build.

**Comment by paulhfu** [comment]
@clbarnes 
Thanks for the comments, i applied your suggested changes.

File: src/python/module/z5py/group.py [hunk 2]
```diff
@@ -228,3 +228,39 @@ def require_dataset(self, name, shape,
...
+        Example:
+
+        # Get a list of all datasets in the file
+        >>> mylist = []
+        >>> def func(name, obj):
+        ...     if isinstance(obj, Dataset):
+        ...         mylist.append(name)
+        ...
+        >>> f = File('foo.n5')
+        >>> f.visititems(func,'')
```
Comment (constantinpape) [inline src/python/module/z5py/group.py]: Please change the function signature in the example as well.

File: src/python/module/z5py/group.py [hunk 3]
```diff
@@ -228,3 +228,39 @@ def require_dataset(self, name, shape,
...
+        """
+        for item in self.items():
+            if isinstance(item[1],Group):
+                func_ret = item[1].visititems(func)
+                if func_ret is not None:
+                    return func_ret
+            else:
+                func_ret = func(os.path.join(self.path, item[0]), self[item[0]])
```
Comment (constantinpape) [inline src/python/module/z5py/group.py]: The second argument of `func` should be `item[1]`. `self[item[0]]` will give the same object, but
is more complicated and obfuscates the code.

Comment (paulhfu) [reply src/python/module/z5py/group.py]: hmm okay, so change it to `item[1]. self[item[0]]` or not?

Comment (paulhfu) [reply src/python/module/z5py/group.py]: ok i understood it now

File: src/python/module/z5py/group.py [hunk 4]
```diff
@@ -228,3 +228,39 @@ def require_dataset(self, name, shape,
...
+        # Get a list of all datasets in the file
+        >>> mylist = []
+        >>> def func(name, obj):
+        ...     if isinstance(obj, Dataset):
+        ...         mylist.append(name)
+        ...
+        >>> f = File('foo.n5')
+        >>> f.visititems(func)
+        """
+        for item in self.items():
```
Comment (clbarnes) [inline src/python/module/z5py/group.py]: Really minor style point - you can unpack `item` here to make its usage more clear later:

    for name, obj in self.items():
        if isinstance(obj, Group):
            # etc

**Comment by constantinpape** [comment]
@paulhfu Thanks for this contribution.

**Pull request merged**